### PR TITLE
Fix pixel CPE position

### PR DESCRIPTION
**Pull request by fwyzard**
#### PR description:

Rewrite the pixel CPE position calculation to improve the reproducibility of the results.

Fix the includes used in `DataFormats/TrackingRecHitSoA` to avoid the dependency on `RecoLocalTracker/SiPixelRecHits`.

Thanks to @missirol for spotitng the source of the discrepancy.

#### PR validation:

Tested on top of CMSSW 14.0.11 running the HLT over 100k events:
  - the CPU-only results are reproducible, with and without these changes;
  - the GPU vs CPU results are in general slightly more reproducible; 
  - the biggest improvement is fixing the reproducibility of `HLT_SingleEle8_v7`, going from 2% of the accepted events to being fully reproducible.

##### GPU vs CPU results, without these changes
  - +4/-5 events in `HLT_SingleEle8_v7` (2% of the accepted events)
  -  +/-4 events in  1 path
  -  +/-3 events in  2 paths
  -  +/-2 events in  6 paths
  -  +/-1 event  in 59 paths

##### GPU vs CPU results, with these changes
  -  +/-3 events in  2 paths
  -  +/-2 events in 12 paths
  -  +/-1 event  in 46 paths

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

To be backported to 14.0.x for data taking.

### CODE REVIEW

**Comment by fwyzard** [comment]
@VinInn does this looks reasonable to you ?

File: RecoLocalTracker/SiPixelRecHits/interface/pixelCPEforDevice.h [hunk 1]
```diff
@@ -234,17 +234,30 @@ namespace pixelCPEforDevice {
     if (cp.minCol[ic] == 0 || cp.maxCol[ic] == uint32_t(detParams.nCols - 1))
       cp.ysize[ic] = -cp.ysize[ic];
 
+    /* original code, kept for documentation
+
```
Comment (VinInn) [inline RecoLocalTracker/SiPixelRecHits/interface/pixelCPEforDevice.h:238]: then delete it!

**Comment by fwyzard** [comment]
test parameters:
  - enable = gpu
  - workflows_gpu = 12634.402,12634.403,12634.404

**Comment by fwyzard** [comment]
please test

File: RecoLocalTracker/SiPixelRecHits/interface/pixelCPEforDevice.h (hunk 1 above)
Comment (VinInn) [reply RecoLocalTracker/SiPixelRecHits/interface/pixelCPEforDevice.h:238]: maybe add "make sure optimization does not produce different code on different architectures "
(or similar)

File: RecoLocalTracker/SiPixelRecHits/interface/pixelCPEforDevice.h [hunk 2]
```diff
@@ -234,17 +234,30 @@ namespace pixelCPEforDevice {
...
     // apply the lorentz offset correction
     float xoff = 0.5f * float(detParams.nRows) * comParams.thePitchX;
```
Comment (slava77) [inline RecoLocalTracker/SiPixelRecHits/interface/pixelCPEforDevice.h:240]: was the original comment actually correct? Isn't only the old L245 with `shiftX` the LA offset, while the rest is a translation wrt center?

Comment (fwyzard) [reply RecoLocalTracker/SiPixelRecHits/interface/pixelCPEforDevice.h:240]: I have no idea, but if can suggest a better description I'm happy to fix it

File: RecoLocalTracker/SiPixelRecHits/interface/pixelCPEforDevice.h [hunk 3]
```diff
@@ -234,17 +234,30 @@ namespace pixelCPEforDevice {
...
     // apply the lorentz offset correction
     auto xPos = detParams.shiftX + (comParams.thePitchX * 0.5f * float(mx)) - xoff;
     auto yPos = detParams.shiftY + (comParams.thePitchY * 0.5f * float(my)) - yoff;
+    */
+
+    /* code rearranged to simplify redundant operations and use explicit FMA instructions to improve reproducibility */
+
+    float xPos = std::fmaf(0.5f * ((float)mx - (float)detParams.nRows) - TrackerTraits::bigPixXCorrection,
+                           comParams.thePitchX,
+                           detParams.shiftX);
```
Comment (slava77) [inline RecoLocalTracker/SiPixelRecHits/interface/pixelCPEforDevice.h:256]: to counter the comment above, I feel like the new code is a bit more readable ;)
There is an extra effort to parse from `fmaf` to what it is in plain arithmetic operations

File: RecoLocalTracker/SiPixelRecHits/interface/pixelCPEforDevice.h (hunk 2 above)
Comment (slava77) [reply RecoLocalTracker/SiPixelRecHits/interface/pixelCPEforDevice.h:240]: I'm not sure of the full provenance/authorship of the comments.

@ferencek please check to see if the comments are descriptive. Thanks.

Comment (fwyzard) [reply RecoLocalTracker/SiPixelRecHits/interface/pixelCPEforDevice.h:240]: the first commit of the original CUDA code (circa 2018) had
```c++
  // apply the lorentz offset correction
  auto xPos = detParams.shiftX + comParams.thePitchX*(0.5f*float(mx)+float(phase1PixelTopology::xOffset));
  auto yPos = detParams.shiftY + comParams.thePitchY*(0.5f*float(my)+float(phase1PixelTopology::yOffset));
```

The other corrections and comments were added over time.

**Comment by jfernan2** [comment]
+1

File: RecoLocalTracker/SiPixelRecHits/interface/pixelCPEforDevice.h (hunk 3 above)
Comment (fwyzard) [reply RecoLocalTracker/SiPixelRecHits/interface/pixelCPEforDevice.h:256]: Would this be a good description of the code ?
```c++
// compute the position relative to the center of the module, taking into account the corrections for the Phase 1 big pixels, the pitch, and the Lorentz shift
```

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
+heterogeneous

**Comment by fwyzard** [comment]
+heterogeneous

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
Sorry @jfernan2, could you sign again ?
I think I only updated the comments since you signed earlier.
Thanks !

**Comment by jfernan2** [comment]
+1

**Comment by rappoccio** [comment]
+1

**Pull request merged**